pkg/dynamicconfig/table.go

```diff
@@ -0,0 +1,136 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package dynamicconfig
+
+import (
+	"sort"
+	"strconv"
+
+	"github.com/cilium/hive/job"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/index"
+
+	"github.com/cilium/cilium/pkg/k8s"
+)
+
+const TableName = "cilium-configs"
+
+var (
+	keyIndex = statedb.Index[DynamicConfig, Key]{
+		Name: "config-name-source-key",
+		FromObject: func(t DynamicConfig) index.KeySet {
+			return index.NewKeySet(index.Stringer(t.Key))
+		},
+		FromKey: index.Stringer[Key],
+		Unique:  true,
+	}
+
+	ByKeyIndex = keyIndex.Query
+
+	keyNameIndex = statedb.Index[DynamicConfig, string]{
+		Name: "config-entry-key",
+		FromObject: func(t DynamicConfig) index.KeySet {
+			return index.NewKeySet(index.String(t.Key.Name))
+		},
+		FromKey: index.String,
+		Unique:  false,
+	}
+	ByKeyNameIndex = keyNameIndex.Query
+)
+
+type Key struct {
+	Name   string
+	Source string
+}
+
+func (k Key) String() string {
+	return k.Name + "/" + k.Source
+}
+
+type DynamicConfig struct {
+	Key   Key
+	Value string
+}
+
+func (d DynamicConfig) TableHeader() []string {
+	return []string{
+		"Key",
+		"Source",
+		"Priority",
+		"Value",
+	}
+}
+
+func (d DynamicConfig) TableRow() []string {
+	return []string{
+		d.Key.Name,
+		d.Key.Source,
+		strconv.Itoa(priorities[d.Key.Source]),
+		d.Value,
+	}
+}
+
+func NewConfigTable(db *statedb.DB) (statedb.RWTable[DynamicConfig], error) {
+	tbl, err := statedb.NewTable(
+		TableName,
+		keyIndex,
+		keyNameIndex,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return tbl, db.RegisterTable(tbl)
+}
+
+func RegisterConfigMapReflector(jobGroup job.Group, db *statedb.DB, rcs []k8s.ReflectorConfig[DynamicConfig], c config) error {
+	if !c.EnableDynamicConfig {
+		return nil
+	}
+	for _, rc := range rcs {
+		if err := k8s.RegisterReflector[DynamicConfig](jobGroup, db, rc); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func GetKey(txn statedb.ReadTxn, table statedb.Table[DynamicConfig], key string) (DynamicConfig, bool) {
+	entries := iterToList(table.List(txn, ByKeyNameIndex(key)))
+
+	if len(entries) == 0 {
+		return DynamicConfig{}, false
+	}
+
+	sortByPriority(entries)
+
+	return entries[0], true
+}
+
+func WatchKey(txn statedb.ReadTxn, table statedb.Table[DynamicConfig], key string) (DynamicConfig, <-chan struct{}) {
+	iter, w := table.ListWatch(txn, ByKeyNameIndex(key))
+	entries := iterToList(iter)
+
+	if len(entries) == 0 {
+		return DynamicConfig{}, w
+	}
+
+	sortByPriority(entries)
+
+	return entries[0], w
+}
+
+func sortByPriority(entries []DynamicConfig) {
+	sort.Slice(entries, func(i, j int) bool {
+		return priorities[entries[i].Key.Source] > priorities[entries[j].Key.Source]
```

Comment:
Question:  not sure if this would actually be meaningful, could we ever expect to see default (i.e. 0) priorities, in such a case the ordering would be undefined.  Would it make sense to brake ties using the key alphabetically?

Reply:
There shouldn't be a case where the priorities are equal considering the source was added properly. Sorting alphabetically would just add more confusion in the code IMO.  Each time a source is added, its corresponding priority needs to be added [here](https://github.com/cilium/cilium/pull/34383/files#diff-e93664a938a87894058f8d6c041422e609d6d271c8828b127a12226bcf07e00dR22)

Reply:
sounds good thanks for the clarification